Sarah room and repo

SarahPanther opened the pull request:
Data-layer addition:

- BreatheRoomDatabase - our single database for the app; contains the Inhaler_Usage_Event_table and the Wearable_Data_table
- BreatheRepository - clean interface between app and data
- BreatheDAO - data access object for our database
- InhalerUsageEvent entity - represents a single, uniquely occuring Inhaler Usage Event 
- WearableData entity - represents a single, uniquely occurring cluster of data gathered from the smart pin

Included some simple tests.

SarahPanther committed:
Added Room Architecture Component libraries to Gradle files

SarahPanther committed:
Created WeatherData class, which encapsulates the area weather conditions (from weather API) at the time of the IUE -> to be contained by IUE class

SarahPanther committed:
Created DiaryEntry class, which encapsulates user entered comments and tags for a given IUE.  Added some comments to the WeatherData class

SarahPanther committed:
Created the WearableData class, which encapsulates data collected from the user's smart pin/wearable and is contained in the IUE class. Changed message variable in DiaryEntry to private.

SarahPanther committed:
Created IUE entity class for Room database.  Reformatted comments in WeatherData.java, DiaryEntry.java, and WearableData.java.  Code compiles and deploys successfully to my Android phone.

SarahPanther committed:
Added Entity and Embedded annotations to IUE class. Renamed variables in WeatherData to have the prefix weather and the one data field in WearableData to have the prefix wearable.

SarahPanther committed:
Added minimal Dao for IUE's with insert and select all

SarahPanther committed:
Forgot to add file for IueDao - added IueDao.java

SarahPanther committed:
Created a minimal IueRoomDatabase - will need to implement a migration strategy.

SarahPanther committed:
Created the IueRepository

SarahPanther committed:
Made a deleteAll method in the IueDao to delete all IUEs in the IUE_table; IMPORTANT -> this should only be called during testing, not in the app

SarahPanther committed:
Fixed Embedded variable fields that couldn't be understood in database (Level, Tag, OffsetDateTime) by adding Converters.java and adding a TypeConverter annotation to the IueRoomDatabase.  Will need to implement conversion from string to OffsetDateTime for API versions < 26 (Oreo).  Changed DiaryEntry to only have 1 tag instead of a list; if we really want a list of tags (multiple tags per DiaryEntry), we will need to add a converter for it to be stored in the IueRoom Database.  Code currently compiles and runs on my Android phone.

SarahPanther committed:
Changed getAllIUEs() to return all IUEs in db in ascending lexographical order based on UTC_ISO_8601_date_time string (this is chronological order). Added a trivial read/write test for the IueRoomDatabase and IUE entity.

SarahPanther committed:
Added a comment TODO to query between dates for the IueDao

SarahPanther committed:
Updated build.gradle and changed to androidx

SarahPanther committed:
Renamed IueDao, IueRepository, IueRoomDatabase and IUE to BreatheDao, BreatheRepository, BreatheRoomDatabase, and InhalerUsageEvent

SarahPanther committed:
Changed table name from IUE_table to InhalerUsageEvent_table

SarahPanther committed:
Added updateInhalerUsageEvent method and query between dates method to the BreatheDao

SarahPanther committed:
Added tests and testing dependencies; renamed some instances that still had IUE/Iue for some reason to InhalerUsageEvent; made WearableData an entity and gave it its own table

SarahPanther committed:
Changed comment for getting rid of AsyncTask in BreatheRepository; forgot to mention AsyncTask insert was changed to a concurrent ExecutorService in the BreatheRepository for the previous commit

SarahPanther committed:
Added some comments where they were needed

SarahPanther committed:
Merge branch 'breath-app' into sarah-Room-and-repo

Youssef-Beltagy commented on Breathe/app/build.gradle:
```diff
@@ -20,15 +20,46 @@ android {
+    // Android Architecture Component libraries
 
+    // Room components
+    implementation "android.arch.persistence.room:runtime:$rootProject.roomVersion"
```
Should this be androidx?

SarahPanther replied:
Maybe, I will test replacing this with AndroidX when I work on LiveData.

Youssef-Beltagy commented on Breathe/.idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -0,0 +1,15 @@
+<component name="InspectionProjectProfileManager">
```
What does this file do?

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java:
```diff
@@ -0,0 +1,89 @@
+    @NonNull // this can never be null
+    // name of the column in Room that stores timeStamp
+    @ColumnInfo(name = "Inhaler_Usage_Event_UTC_ISO_8601_date_time")
+    private OffsetDateTime inhalerUsageEventTimeStamp;
```
Is it worthwhile to change this to instant?

SarahPanther replied:
Yes, thank you for changing this to instant.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -0,0 +1,44 @@
+    // Returns all InhalerUsageEvents from the InhalerUsageEvent_table in lexographical
+    // ascending order of string timestamp (because of UTC format, this is in chronological order)
+    @Query("SELECT * FROM InhalerUsageEvent_table " +
+            "ORDER BY Inhaler_Usage_Event_UTC_ISO_8601_date_time ASC")
```
Might Be better to get DESC

SarahPanther replied:
Changed to DESC

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java:
```diff
@@ -0,0 +1,45 @@
+    private static volatile BreatheRoomDatabase INSTANCE; // this BreatheRoomDatabase is a singleton
+    private static final int NUMBER_OF_THREADS = 4;
+    // Executor Service is used (replaced AsyncTask) so inserts may happen concurrently
+    static final ExecutorService dbWriteExecutor = Executors.newFixedThreadPool(NUMBER_OF_THREADS);
```
Why use a fixed size thread pool?

Does it have any effect on a component's lifecycle?

SarahPanther replied:
changed to cachedThreadPool

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/WeatherData.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * 3 state level (used for pollen index)
+ */
+enum Level {
```
Can we rename this to PollenLevel?

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/InhalerUsageEvent.java:
```diff
@@ -0,0 +1,89 @@
+    // I made these public because all of the Embedded examples I saw were public
+    // For more info on Embedded objects within Entities,
+    // see https://developer.android.com/training/data-storage/room/relationships
+    @Embedded public WeatherData weatherData;
```
You should be able to make it private as long as you make setters/getters.

SarahPanther replied:
Changed to private

Youssef-Beltagy commented on .idea/Smart-Inhaler_Room.iml:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
You must delete this file.

SarahPanther replied:
Deleted

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheDao.java:
```diff
@@ -0,0 +1,44 @@
+
+
+    // Return InhalerUsageEvents between the parameter dates (inclusive)
+    @Query("SELECT * FROM InhalerUsageEvent_table " +
```
String comparisons use length. So you need to make sure all the timestamps are the same lenght.

Youssef-Beltagy commented on Breathe/app/src/main/java/com/ybeltagy/breathe/BreatheRoomDatabase.java:
```diff
@@ -0,0 +1,45 @@
+    public abstract BreatheDao breatheDao();
+
+    private static volatile BreatheRoomDatabase INSTANCE; // this BreatheRoomDatabase is a singleton
+    private static final int NUMBER_OF_THREADS = 4;
```
Don't use fixed size thread pools.

SarahPanther replied:
Changed to cachedThreadPool

Youssef-Beltagy committed:
Instant, WearableData, and cachedThreadPool

I modified the code to use Instant instead of OffsetDateTime.

I modified WearableData class so it contains the data the wearable is expected to send.

I modified the repository to use a cachedThreadPool.

I ran the instrumented tests and they were successful.

Youssef-Beltagy committed:
Merge branch 'sarah-Room-and-repo' into room-and-repo-review

Youssef-Beltagy committed:
Removed useless import statement.

SarahPanther committed:
Merge pull request #15 from Youssef-Beltagy/room-and-repo-review

Instant, WearableData, and cachedThreadPool

SarahPanther commented:
- Left unresolved comments so we can keep them in mind as we move forward
- Resolved comments on things Youssef fixed in his review branch of this branch
